Issue compiling Freeorion on Fedora 32

* **FreeOrion Version**: GitHub Master from 3/18/20
* **Operating System**: Fedora 32


### Description

While running make I receive the following error:

````
[ 16%] Built target GiGi
Scanning dependencies of target freeorionversion
Writing file: util/Version.cpp
Building v0.4.9+ build 2020-03-18.f835296
[ 16%] Built target freeorionversion
Scanning dependencies of target freeorioncommon
[ 17%] Building CXX object CMakeFiles/freeorioncommon.dir/combat/CombatEvent.cpp.o
In file included from /usr/include/boost/serialization/deque.hpp:25,
                 from /home/dahlgren/freeorion/combat/../util/Serialize.ipp:13,
                 from /home/dahlgren/freeorion/combat/CombatEvent.cpp:6:
/usr/include/boost/serialization/collections_save_imp.hpp: In function ‘void boost::serialization::stl::save_collection(Archive&, const Container&, boost::serialization::collection_size_type)’:
/usr/include/boost/serialization/collections_save_imp.hpp:60:16: error: no post-decrement operator for type
   60 |     while(count-- > 0){
      |                ^~
In file included from /usr/include/boost/serialization/deque.hpp:26,
                 from /home/dahlgren/freeorion/combat/../util/Serialize.ipp:13,
                 from /home/dahlgren/freeorion/combat/CombatEvent.cpp:6:
/usr/include/boost/serialization/collections_load_imp.hpp: In function ‘typename boost::enable_if<boost::serialization::detail::is_default_constructible<typename T::value_type>, void>::type boost::serialization::stl::collection_load_impl(Archive&, T&, boost::serialization::collection_size_type, boost::serialization::item_version_type)’:
/usr/include/boost/serialization/collections_load_imp.hpp:72:16: error: no post-decrement operator for type
   72 |     while(count-- > 0){
      |                ^~
/usr/include/boost/serialization/collections_load_imp.hpp: In function ‘typename boost::disable_if<boost::serialization::detail::is_default_constructible<typename T::value_type>, void>::type boost::serialization::stl::collection_load_impl(Archive&, T&, boost::serialization::collection_size_type, boost::serialization::item_version_type)’:
/usr/include/boost/serialization/collections_load_imp.hpp:94:16: error: no post-decrement operator for type
   94 |     while(count-- > 0){
      |                ^~
In file included from /home/dahlgren/freeorion/combat/../util/Serialize.ipp:15,
                 from /home/dahlgren/freeorion/combat/CombatEvent.cpp:6:
/usr/include/boost/serialization/map.hpp: In function ‘void boost::serialization::load_map_collection(Archive&, Container&)’:
/usr/include/boost/serialization/map.hpp:58:16: error: no post-decrement operator for type
   58 |     while(count-- > 0){
      |                ^~
In file included from /usr/include/boost/serialization/unordered_set.hpp:24,
                 from /home/dahlgren/freeorion/combat/../util/Serialize.ipp:16,
                 from /home/dahlgren/freeorion/combat/CombatEvent.cpp:6:
/usr/include/boost/serialization/unordered_collections_save_imp.hpp: In function ‘void boost::serialization::stl::save_unordered_collection(Archive&, const Container&)’:
/usr/include/boost/serialization/unordered_collections_save_imp.hpp:69:16: error: no post-decrement operator for type
   69 |     while(count-- > 0){
      |                ^~
In file included from /usr/include/boost/serialization/unordered_set.hpp:25,
                 from /home/dahlgren/freeorion/combat/../util/Serialize.ipp:16,
                 from /home/dahlgren/freeorion/combat/CombatEvent.cpp:6:
/usr/include/boost/serialization/unordered_collections_load_imp.hpp: In function ‘void boost::serialization::stl::load_unordered_collection(Archive&, Container&)’:
/usr/include/boost/serialization/unordered_collections_load_imp.hpp:64:16: error: no post-decrement operator for type
   64 |     while(count-- > 0){
      |                ^~
In file included from /home/dahlgren/freeorion/combat/../util/Serialize.ipp:18,
                 from /home/dahlgren/freeorion/combat/CombatEvent.cpp:6:
/usr/include/boost/serialization/set.hpp: In function ‘void boost::serialization::load_set_collection(Archive&, Container&)’:
/usr/include/boost/serialization/set.hpp:53:16: error: no post-decrement operator for type
   53 |     while(count-- > 0){
      |                ^~
In file included from /home/dahlgren/freeorion/combat/../util/Serialize.ipp:19,
                 from /home/dahlgren/freeorion/combat/CombatEvent.cpp:6:
/usr/include/boost/serialization/vector.hpp: In function ‘void boost::serialization::save(Archive&, const std::vector<bool, Allocator>&, unsigned int)’:
/usr/include/boost/serialization/vector.hpp:191:16: error: no post-decrement operator for type
  191 |     while(count-- > 0){
      |                ^~
/usr/include/boost/serialization/vector.hpp: In function ‘void boost::serialization::load(Archive&, std::vector<bool, Allocator>&, unsigned int)’:
/usr/include/boost/serialization/vector.hpp:207:72: error: no pre-increment operator for type
  207 |     for(collection_size_type i = collection_size_type(0); i < count; ++i){
      |                                                                        ^
make[2]: *** [CMakeFiles/freeorioncommon.dir/build.make:63: CMakeFiles/freeorioncommon.dir/combat/CombatEvent.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:631: CMakeFiles/freeorioncommon.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
````

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -212,6 +212,8 @@ if(UNIX AND NOT APPLE)
     add_definitions(-DENABLE_BINRELOC -DBINPATH=\"${CMAKE_INSTALL_FULL_BINDIR}\" -DSHAREPATH=\"${CMAKE_INSTALL_FULL_DATADIR}\")
 endif()
 
+include(TestForBug2795)
+
 
 ##
 ## Collect project dependencies.

--- a/cmake/TestForBug2795.cmake
+++ b/cmake/TestForBug2795.cmake
@@ -1,0 +1,53 @@
+# This file is a CMake configuration source compile check
+#
+# G++ 10.0.1 had a regression bug, which causes boost::serialization to fail
+# to compile because the underlying type of a strong typedef template did not
+# expose its ++/-- post operators.
+#
+# For more details see:
+#   https://github.com/freeorion/freeorion/issues/2795
+#   https://github.com/boostorg/serialization/issues/192
+#   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94190
+
+
+include(CheckCXXSourceCompiles)
+
+# Enforce compiler to be invoked with C locale to ensure non-localized
+# error messages.
+set(_locale_vars LC_ALL LC_MESSAGES LANG)
+foreach(v IN LISTS _locale_vars)
+    set(_locale_vars_saved_${v} "$ENV{${v}}")
+    set(ENV{${v}} C)
+endforeach()
+
+check_cxx_source_compiles("
+struct A
+{
+    int _i;
+    operator int&() { return _i; }
+};
+
+template <typename T> void f()
+{
+    A a;
+    a++; // triggers bug
+}
+
+int main() { f<int>(); }
+" NOT_GNU_CXX_BUG_94190 FAIL_REGEX "no post-decrement operator for type")
+
+if(NOT NOT_GNU_CXX_BUG_94190)
+    message(FATAL_ERROR
+        "The found version of the g++ compiler suffers from bug"
+        " https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94190"
+        " which triggers a known bug in combination with boost::serialization."
+        " Please update the compiler to a more recent version of g++ or"
+        " use an alternative compiler."
+        " After update rebuild the project from a clean build tree."
+    )
+endif()
+
+# Restore locale settings
+foreach(v IN LISTS _locale_vars)
+    set(ENV{${v}} ${_locale_vars_saved_${v}})
+endforeach()

